sbin/clri/clri.8
.\"	$OpenBSD: clri.8,v 1.12 2007/05/31 19:19:43 jmc Exp $
.\"	$NetBSD: clri.8,v 1.2 1995/03/18 14:54:31 cgd Exp $
.\"
.\" Copyright (c) 1980, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)clri.8	8.2 (Berkeley) 4/19/94
.\"
.Dd $Mdocdate: May 31 2007 $
.Dt CLRI 8
.Os
.Sh NAME
.Nm clri
.Nd clear inodes
.Sh SYNOPSIS
.Nm clri
.Ar special_device inode_number Op Ar ...
.Sh DESCRIPTION
.Bf -symbolic
.Nm
is obsoleted for normal file system repair work by
.Xr fsck 8 .
.Ef
.Pp
.Nm
zeros out the inodes with the specified inode numbers
on the filesystem residing on the given
.Ar special_device .
The
.Xr fsck 8
utility is usually run after
.Nm
to reclaim the zero'ed inodes and the
blocks previously claimed by those inodes.
Both read and write permission are required on the specified
.Ar special_device .
.Pp
The primary purpose of this routine
is to remove a file which
for some reason is not being properly handled by
.Xr fsck 8 .
Once removed,
it is anticipated that
.Xr fsck 8
will be able to clean up the resulting mess.
.Sh SEE ALSO
.Xr inode 5 ,
.Xr fsck 8 ,
.Xr fsdb 8 ,
.Xr ncheck 8
.Sh BUGS
If the file is open, the work of
.Nm
will be lost when the inode is written back to disk from the inode cache.


sbin/clri/clri.c
/*	$OpenBSD: clri.c,v 1.11 2007/05/21 18:13:11 millert Exp $	*/
/*	$NetBSD: clri.c,v 1.19 2005/01/20 15:50:47 xtraeme Exp $	*/

/*
 * Copyright (c) 1990, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Rich $alz of BBN Inc.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if 0
#ifndef lint
static const char copyright[] =
"@(#) Copyright (c) 1990, 1993\n\
	The Regents of the University of California.  All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)clri.c	8.2 (Berkeley) 9/23/93";
#endif /* not lint */
#endif

#include <sys/param.h>

#include <ufs/ufs/dinode.h>
#include <ufs/ffs/fs.h>

#include <err.h>
#include <fcntl.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <unistd.h>

/*
 * Possible superblock locations ordered from most to least likely.
 */
static int sblock_try[] = SBLOCKSEARCH;

static void
usage(void)
{
	(void)fprintf(stderr, "usage: clri special_device inode_number ...\n");
	exit(1);
}

int
main(int argc, char *argv[])
{
	struct ufs1_dinode *dp1;
	struct ufs2_dinode *dp2;
	struct fs *sbp;
	char *ibuf[MAXBSIZE];
	char *fs, sblock[SBLOCKSIZE];
	size_t bsize;
	off_t offset;
	int i, fd, imax, inonum;

	if (argc < 3)
		usage();

	fs = *++argv;
	sbp = NULL;

	/* get the superblock. */
	if ((fd = open(fs, O_RDWR, 0)) < 0)
		err(1, "%s", fs);
	for (i = 0; sblock_try[i] != -1; i++) {
		offset = (off_t)(sblock_try[i]);
		if (pread(fd, sblock, sizeof(sblock), offset) != sizeof(sblock))
			err(1, "%s: can't read superblock", fs);
		sbp = (struct fs *)sblock;
		if ((sbp->fs_magic == FS_UFS1_MAGIC ||
		     (sbp->fs_magic == FS_UFS2_MAGIC &&
		      sbp->fs_sblockloc == sblock_try[i])) &&
		    sbp->fs_bsize <= MAXBSIZE &&
		    sbp->fs_bsize >= (int)sizeof(struct fs))
			break;
	}
	if (sblock_try[i] == -1)
		errx(2, "cannot find file system superblock");
	bsize = sbp->fs_bsize;

	/* check that inode numbers are valid */
	imax = sbp->fs_ncg * sbp->fs_ipg;
	for (i = 1; i < (argc - 1); i++) {
		if (atoi(argv[i]) <= 0 || atoi(argv[i]) >= imax) 
			errx(1, "%s is not a valid inode number", argv[i]);
	}

	/* clear the clean flag in the superblock */
	if (sbp->fs_inodefmt >= FS_44INODEFMT) {
		sbp->fs_clean = 0;
		if (pwrite(fd, sblock, sizeof(sblock), offset) != sizeof(sblock))
			err(1, "%s: can't update superblock", fs);
		(void)fsync(fd);
	}

	/* remaining arguments are inode numbers. */
	while (*++argv) {
		/* get the inode number. */
		inonum = atoi(*argv);
		(void)printf("clearing %d\n", inonum);

		/* read in the appropriate block. */
		offset = ino_to_fsba(sbp, inonum);	/* inode to fs blk */
		offset = fsbtodb(sbp, offset);		/* fs blk disk blk */
		offset *= DEV_BSIZE;			/* disk blk to bytes */

		/* seek and read the block */
		if (pread(fd, ibuf, bsize, offset) != bsize)
			err(1, "%s", fs);

		if (sbp->fs_magic == FS_UFS2_MAGIC) {
			/* get the inode within the block. */
			dp2 = &(((struct ufs2_dinode *)ibuf)
			    [ino_to_fsbo(sbp, inonum)]);

			/* clear the inode, and bump the generation count. */
			memset(dp2, 0, sizeof(*dp2));
			dp2->di_gen = arc4random();
		} else {
			/* get the inode within the block. */
			dp1 = &(((struct ufs1_dinode *)ibuf)
			    [ino_to_fsbo(sbp, inonum)]);

			/* clear the inode, and bump the generation count. */
			memset(dp1, 0, sizeof(*dp1));
			dp1->di_gen = arc4random();
		}

		/* backup and write the block */
		if (pwrite(fd, ibuf, bsize, offset) != bsize)
			err(1, "%s", fs);
		(void)fsync(fd);
	}
	(void)close(fd);
	exit(0);
}
